### PR TITLE
Basic workflow controller

### PR DESCRIPTION
**Pull request by cd1989**
**What this PR does / why we need it**:

Add your description

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @zhujian7 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/workflow/controller/handlers/workflowrun/pod.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+	// Add init containers for each input resource and also bind resource to workload containers.
+	for _, r := range stage.Spec.Pod.Inputs.Resources {
+		resource, err := m.client.CycloneV1alpha1().Resources(m.wfr.Namespace).Get(r.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		var image string
+		switch resource.Spec.Type {
```
Comment (supereagle) [inline pkg/workflow/controller/handlers/workflowrun/pod.go]: Use resolver map.

File: pkg/workflow/controller/handlers/workflowrun/pod.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+		for _, p := range resource.Spec.Parameters {
+			envs = append(envs, corev1.EnvVar{
+				Name:  p.Name,
+				Value: p.Value,
+			})
+		}
+		container := corev1.Container{
+			Name:  r.Name,
+			Image: image,
+			Args:  []string{"pull"},
```
Comment (supereagle) [inline pkg/workflow/controller/handlers/workflowrun/pod.go]: Define `pull` as a const.

File: pkg/workflow/controller/handlers/workflowrun/status.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	_, ok := wfr.Status.Stages[stage]
+	if !ok {
+		wfr.Status.Stages[stage] = &v1alpha1.StageStatus{}
+	}
+	wfr.Status.Stages[stage].Status = status
+}
+
+func OverallStatus(wfr *v1alpha1.WorkflowRun) string {
+	var running, waiting, err bool
+	for _, stage := range wfr.Status.Stages {
```
Comment (supereagle) [inline pkg/workflow/controller/handlers/workflowrun/status.go]: If there are parallel stages, how to determine the overall status?

File: pkg/workflow/controller/config.go [hunk 4]
```diff
@@ -10,10 +10,21 @@ import (
...
+const (
+	// The key of config file in ConfigMap
+	ConfigFileKey = "workflow-controller.json"
+
+	// Keys of resources images and coordinator image in config file
+	GitResolverImage   = "git-resolver"
+	ImageResolverImage = "image-resolver"
+	KvResolverImage    = "kv-resolver"
+	CoordinatorImage   = "coordinator"
+)
```
[the remark this answers is not in the document]
Comment (cd1989) [reply pkg/workflow/controller/config.go]: Done

**Pull request merged**